### PR TITLE
Better API for ScriptWithExtensions

### PR DESCRIPTION
**Pull request by Manishearth**
This:

 - Moves the datapayload type to icu::properties::provider, and merges it with ScriptWithExtensionsPropertyV1
   - n.b. this is done in stages, the intermediate stages may be a bit redundant
 - (re)Adds a ScriptWithExtensions/ScriptWithExtensionsBorrowed type pair and gives them all the methods and docs

todo: FFI

### CODE REVIEW

File: components/properties/src/provider.rs [hunk 1]
```diff
@@ -67,9 +70,32 @@ pub enum PropertyCodePointMapV1<'data, T: TrieValue> {
...
+    /// | 3                       | First value in sub-array, index given by lower 10 bits | Sub-array excluding first value, index given by lower 10 bits  |
+    /// | 2                       | Script=Inherited                                       | Entire sub-array, index given by lower 10 bits                 |
+    /// | 1                       | Script=Common                                          | Entire sub-array, index given by lower 10 bits                 |
+    /// | 0                       | Value in lower 10 bits                                 | `[ Script value ]` single-element array                        |
+    ///
+    /// When the lower 10 bits of the value are used as an index, that index is
+    /// used for the outer-level vector of the nested `extensions` structure.
     #[cfg_attr(feature = "serde", serde(borrow))]
-    pub data: ScriptWithExtensions<'data>,
+    #[doc(hidden)] // #2417
```
Comment (sffc) [inline components/properties/src/provider.rs:88]: Nit: We said that we don't hide the fields of provider structs, only library types

File: components/properties/src/provider.rs [hunk 2]
```diff
@@ -176,6 +202,125 @@ impl<'data, T: TrieValue> PropertyCodePointMapV1<'data, T> {
     }
 }
 
+// See ScriptWithExtensions for documentation of these functions
```
Comment (sffc) [inline components/properties/src/provider.rs:205]: Nit: Please move these functions to ScriptWithExtensionsBorrowed. There should be very little code in provider.rs files (I know we violate that sometimes but I really would like us to have this separation wherever possible)

File: ffi/diplomat/ffi_coverage/tests/missing_apis.txt [hunk 3]
```diff
@@ -1,12 +1,4 @@
-icu::properties::script::ScriptExtensionsSet#Struct
-icu::properties::script::ScriptExtensionsSet::contains#FnInStruct
 icu::properties::script::ScriptExtensionsSet::iter#FnInStruct
```
Comment (Manishearth) [inline ffi/diplomat/ffi_coverage/tests/missing_apis.txt:1]: @sffc With these changes the only remaining bit is this function. I could create an iterator type, or I could just expose a `/len()` and `.script_at()`, thoughts?

I think the iterator is less nice over FFI until we can have good iterators; but we may potentially change the impl to not be a vector in the future? Though we can just declare that iteration order is unstable.

Comment (Manishearth) [reply ffi/diplomat/ffi_coverage/tests/missing_apis.txt:1]: The annoying thing over exposing len/at is that then the Rust types need to do that, but there is no reason to have this on the Rust side (perhaps I can just doc(hidden) them)

File: components/properties/src/script.rs [hunk 4]
```diff
@@ -177,10 +175,10 @@ impl From<ScriptWithExt> for Script {
 }
 
 /// A struct that wraps a [`Script`] array, such as in the return value for
-/// [`get_script_extensions_val`](ScriptWithExtensions::get_script_extensions_val).
+/// [`get_script_extensions_val`](ScriptWithExtensionsPropertyV1::get_script_extensions_val).
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
 pub struct ScriptExtensionsSet<'a> {
-    values: &'a ZeroSlice<Script>,
+    pub(crate) values: &'a ZeroSlice<Script>,
```
Comment (sffc) [inline components/properties/src/script.rs:181]:
```suggestion
    values: &'a ZeroSlice<Script>,
```

File: components/properties/src/script.rs [hunk 5]
```diff
@@ -7,45 +7,43 @@
...
+use zerovec::{ule::AsULE, ZeroSlice};
 
 /// The number of bits at the low-end of a `ScriptWithExt` value used for
 /// storing the `Script` value (or `extensions` index).
 const SCRIPT_VAL_LENGTH: u16 = 10;
 
 /// The bit mask necessary to retrieve the `Script` value (or `extensions` index)
 /// from a `ScriptWithExt` value.
-const SCRIPT_X_SCRIPT_VAL: u16 = (1 << SCRIPT_VAL_LENGTH) - 1;
+pub(crate) const SCRIPT_X_SCRIPT_VAL: u16 = (1 << SCRIPT_VAL_LENGTH) - 1;
```
Comment (sffc) [inline components/properties/src/script.rs:25]: Nit: You can undo some of these `pub(crate)` now that the code is back in this file

```suggestion
const SCRIPT_X_SCRIPT_VAL: u16 = (1 << SCRIPT_VAL_LENGTH) - 1;
```

File: ffi/diplomat/src/script.rs [hunk 6]
```diff
@@ -32,20 +38,68 @@ pub mod ffi {
...
+        #[diplomat::rust_link(
+            icu::properties::script::ScriptWithExtensions::as_borrowed,
+            FnInStruct
+        )]
+        pub fn as_borrowed<'a>(&'a self) -> Box<ICU4XScriptWithExtensionsBorrowed<'a>> {
+            Box::new(ICU4XScriptWithExtensionsBorrowed(self.0.as_borrowed()))
         }
+    }
 
+    impl<'a> ICU4XScriptWithExtensionsBorrowed<'a> {
```
Comment (sffc) [inline ffi/diplomat/src/script.rs:67]: Thought: We don't expose `CodePointSetDataBorrowed` due to simplicity and I think the assumption that the cost of crossing FFI overshadows the cost of the yoke. Do we need `ScriptWithExtensionsBorrowed` over FFI then?

File: ffi/diplomat/ffi_coverage/tests/missing_apis.txt (hunk 3 above)
Comment (sffc) [reply ffi/diplomat/ffi_coverage/tests/missing_apis.txt:1]: len/at seems fine. We should use that paradigm more often over FFI.

I see pros and cons to making them public vs hidden in Rust; either is fine.

File: ffi/diplomat/src/script.rs (hunk 6 above)
Comment (Manishearth) [reply ffi/diplomat/src/script.rs:67]: We have to for the scx set stuff to work, from yesterday's discussion. That's part of why I wanted to go back and redesign this.

**Review by sffc — APPROVED** [review]

**Pull request merged**